### ISSUE TITLE
Improvement : Add the testID prop to the supported native components

### ISSUE THREAD

**Issue by omatrot** [Microsoft/reactxp#563]
In order to safely use a test framework like [Detox](https://github.com/wix/detox/blob/master/docs/Introduction.WritingFirstTest.md#step-3-add-testids-to-your-app-to-assist-in-matching-elements), we have to rely on the React Native's `testID` prop that is not exposed yet.

**Comment by dryganets** [comment]
It should be easy to implement but we should consider the following things:
1. Don't set any value by default.
2. Make sure that we can strip it in the release.
 
Setting this property is harmful to Android performance as layoutOnly views with this property set can't be removed from the view hierarchy and the [FLAT](https://github.com/facebook/react-native/blob/1e8f3b11027fe0a7514b4fc97d0798d3c64bc895/ReactAndroid/src/main/java/com/facebook/react/flat/README.md) feature if used can't optimize it as well.

**Comment by luisnaranjo733** [comment]
It's worth noting that removing ```testID``` for release will block e2e testing via testID for release builds, which is a valid scenario. It seems like it would be best to leave the removal of this prop up to the client app.
Or at least making the behavior configurable so that apps can build release flavors that don't strip this prop out.

@a-tarasyuk I haven't verified, but it sounds to me like rn does not strip this in release, and leaves it up to the user. Here is the [documentation](https://facebook.github.io/react-native/docs/view.html#testid) for this prop.

**Comment by dryganets** [comment]
The presence of this property in the release will kill performance on Android.

**Comment by luisnaranjo733** [comment]
I totally get that this is undesirable behavior for GA-release. But it is desirable despite the perf hit on e2e-release.

What if we did something like this?

1. Don't set any value by default
2. Document this prop in [components/View](https://microsoft.github.io/reactxp/docs/components/view.html), and add an ominous perf warning ([just like rn does](https://facebook.github.io/react-native/docs/view.html#testid))
3. Strip the prop in release if exists _unless_ the app has explicitly opted-out of this safety check (Maybe via a 3rd optional param to RX.Apps.Initialize or a new opts object)

OR

1. Don't set any value by default (so perf hit will only happen if people explicitly set the prop)
2. Document this prop in [components/View](https://microsoft.github.io/reactxp/docs/components/view.html), and add an ominous perf warning ([just like rn does](https://facebook.github.io/react-native/docs/view.html#testid))
3. Console.warn as extra guard to prevent people from doing this in release

Either of these approaches would ensure that

1. People aren't accidentally killing the perf of their Android GA builds
2. E2E testing on release builds w/ testID is possible

Also, I want to emphasize that the reason this is important is that using anything other than the rn testID prop to find elements makes e2e tests much more brittle. Currently, you have to key on accessibilityLabel or text props for matching. You can imagine how that would fall down easily if the content changes or if you are trying to test something like localization.

**Comment by erictraut** [comment]
A couple of thoughts and questions...

If we add this, we should probably call it "testId" (note the lower-case "d") to maintain capitalization consistency with the rest of ReactXP. We'd need to transform it for compatibility with RN.

How would "testId" get interpreted on the web platform?

Would the testId prop be needed on all RX components or just RX.View? I presume that it would be needed everywhere.

@dryganets, why do you think the addition of testId will block the perf strategies on Android? I presume that only the View elements that have testID props would be ineligible for view hierarchy removal and flattening. I also presume that testID's will typically be provided only for key View elements that are required for test execution, not every View element in the app. These key elements probably already have other props that prevent them from being removed or flattened. For these reasons, I'm not so worried about leaving the testID props in place for release builds. 

Unless/until we have evidence that there is an actual perf impact, I'd prefer to avoid adding logic in ReactXP that strips out the testId prop at runtime. That has its own runtime overhead.

**Comment by luisnaranjo733** [comment]
> How would "testId" get interpreted on the web platform?

The most reliable/least brittle locator strategy for e2e testing on web is to match via the id attribute of a DOM element. This is pretty easy to do for web, since it's an extremely common pattern to use this id prop for styling anyways.

The common [RX.View interface](https://github.com/Microsoft/reactxp/blob/master/src/common/Types.ts#L622) already has this web-only ```id?: string``` prop, and the [web implementation](https://github.com/Microsoft/reactxp/blob/master/src/web/View.tsx#L362) of RX.View passes this id to a div.

[Idea] What if instead of creating a ```testId``` prop we simply re-use this existing web-only ```id``` prop and fill in the mobile implementation to map to rn's ```testID```? This way there is no need to maintain similar ```id``` and ```testId``` variations.

> Would the testId prop be needed on all RX components or just RX.View? I presume that it would be needed everywhere.

That's a great question. I could be wrong on this, but I think all RX components that extend built-in RN components will need to plumb this in. I don't think that plumbing this in to the root View will be enough to get it for free everywhere else. Unlike native Android and iOS views, not all components extend the base View class (this is true in RN and XP as far as I can tell).

For example, if you look at the docs for the [RN Image](https://facebook.github.io/react-native/docs/image.html#testid) component, it specifies that it takes in a testID prop.

**Comment by erictraut** [comment]
I like the idea of combining ```id``` and ```testId```. I wish we wouldn't have added the ```id``` prop in the first place. It's a terribly generic and non-descriptive name for a prop. I'd like to deprecate it over time.

So here's a variation on your idea... We expose a new ```testId``` prop on all RX components. On RN platforms, it is used to set the ```testID``` prop. On the web, it's used to set the ```id``` attribute. For RX.View on the web, we continue to support ```id``` as an alternate way of setting this property, but I document it as "deprecated" and give users a fair warning that they need to switch their usage over to ```testId```. Then I'll completely remove ```id``` in some future release.

**Comment by dryganets** [comment]
@erictraut, my concern is that with big enough code base it is easy to add the testId to not appriopriate places. With time errors like this are tend to accumulate. I will be better to exclude such a problem in production.

We could warn in react-native release for the cases where testId is the only reason for not optimizing the view out. But I believe it is better to have one thing less to think about rather than one more.

As for the flat it is more of the future concern as in react-native version we are using it is not production ready, yet.
The flat optimization is a little bit harder to reason about - It is designed to combine multiple views into the single one. In case you have a testId in one of the components it will lead to the whole other siblings of this component to be not optimized.

**Comment by luisnaranjo733** [comment]
I like it. But is there a better name than ```testId``` for this new prop? Having "test" in the name implies that its utility is restricted to testing scenarios (which is  true for native afaik, not so much for web).  The ```id```  DOM attr is  commonly used in web today for styling and deep linking (#<id> in url). In theory, users could still achieve these things by moving from the deprecated ```id``` to ```testId```, but the name of this new prop doesn't make as much sense on web.

Styling example:

```css
.navbar {
    list-style-type: none;
    margin: 0;
    padding: 0;
}
```

```html
<ul id="navbar">
  <li><a href="/home">Home</a></li>
  <li><a href="/new">News</a></li>
  <li><a href="/contact">Contact</a></li>
  <li><a href="/about">About</a></li>
</ul>
```

Deep linking example:

 Navigating to website.com/#contact will deep link the user to this section
```html
...
<h2 id="contact">Contact us</h2>
...
```

**Comment by erictraut** [comment]
That's a good point.

Here's another idea. We could use the existing ```accessibilityId```. It's currently used only for focus arbitration (i.e. when more than one element wants keyboard autofocus). If accessibilityId is specified, we can pass it through to ```testID``` on RN and ```id``` on web.

@mshoho, you added ```accessibilityId``` recently. What do you think about using it for this purpose?

**Comment by luisnaranjo733** [comment]
Wouldn't that override whatever accessibilityId is used for today?

I've got a similar suggestion - Why not just use the currently web only ```id``` prop to pass through to the RN ```testID``` on mobile? The only concern we had about that approach was the Android perf hit. If we're taking the approach of not worrying about that for now, this could be a good option.

**Comment by erictraut** [comment]
I don't want to perpetuate the use of ```id```. I want to deprecate it.

I think ```accessibilityId``` will work well. Let's see what @mshoho thinks.

**Comment by robeeeert** [comment]
Hi guys, we used to use `data-guitest` props for our GUI tests. With RX.View, this does not work anymore. The approach discussed here is very promising, but the usage of `id` would not be optimal, since in our case many views can have the same `data-guitest` key (e.g. for counting elements).

**Comment by mshoho** [comment]
My main concern about using one for another is about mixing the concepts which might eventually blow up.

From the discussion above, we have three things which might partially intersect but not exactly interchangeable.

The purpose of accessibilityId is to give the application an ability to distinguish one component instance from another from the application code regardless of the environment (if needed and without extra overhead). And I have more use cases for it in mind.

Web's id is a wider thing for web. And I agree that in a form it is now, it should most likely go away from the public API, as we've added it like it is when we were desperately trying to fix some accessibility issues and in the ongoing accessibility API refactoring work I'll move it to the web's accessibility properties container.

Testing is a third thing. And I think it's not a bad idea to have something specifically dedicated for it. Given the performance concerns and the fact that it's sometimes desired to have testId in the production builds too, I see it probably like this:

1. Add testId prop to common props which will go to testID in the React Native environment.
2. On web, when testId is present, add an attribute like data-test-id to the component's DOM node.
3. In RX.UserInterface (or maybe in some better suited place) add a method called something like setTestIdEnabled(enabled: boolean), and make the value false by default. RX should skip setting testID and data-test-id unless it's explicitly enabled.
3a. As an alternative to 3, we can have RX dev build and RX production build (like React itself has). We already have AppConfig.isDevelopmentMode() which is a function call and nooping it in the production build would perform better.

That way the application can have as many testIds as it needs without iffing it for every component but it will have to switch them on if needed.

**Comment by a-tarasyuk** [comment]
@mshoho Based on the last two items in your list, I understood that **ReactXP** has to care about removing `testId` based on environment _(test, prod, dev)_, and it should be done something like this

```typescript
export class AppConfig {
    private _isTest = false;
    ....
    setAppConfig(isDebug: boolean, isDevelopment: boolean, isTest: boolean) {
       ...
        this._isTest = isTest;
    }
    ....
    isTestMode(): boolean {
        return this._isTest;
    }
}
```

```typescript
const buildTestId = (testId: string|undefined): string|undefined => {
   if (AppConfig.isTestMode()) {
       return testId;
   }
   
   return undefined;
}
```

RN
```typescript
<RN.Text testID={ buildTestId(this.prop.testId) } />
```

Web
```typescript
<span data-test-id={ buildTestId(this.prop.testId) } />
```

Is it true?

**Comment by mshoho** [comment]
Yes, something like that. Though the growing amount of the function calls which should be noop in production makes me thinking that having separate builds for the separate environment would be better. But it's also trickier.

**Comment by erictraut** [comment]
OK, you've made good points about not combining these concepts. Let's go with a new prop ```testId``` and expose it as @mshoho suggests above.

I'm not convinced there's any real perf impact here, so I'd like to refrain from adding conditional logic for now. We can always add this later if there's evidence of perf issues. I doubt there ever will be a need for it because 1) these testId's will be used for a small fraction of the elements in the view hierarchy and 2) they will be used almost exclusively for elements that already have props that prevent them from being elided or flattened.

**Comment by erictraut** [comment]
The implementation is straightforward, but it involves touching a bunch of files. I've applied the change  already. I'll push it once I have the docs and tests updated as well.

**Comment by erictraut** [comment]
The change is pushed and published in 1.3.0-rc.6. I've done some testing with it, but let me know if you see any problems.

**Comment by luisnaranjo733** [comment]
Looks great. Thanks!
